diff --git a/src/globals.h b/src/globals.h
index d215066..7f3b46e 100644
--- a/src/globals.h
+++ b/src/globals.h
@@ -639,7 +639,11 @@ struct Global
     /*
      * regex.c
      */
+#if defined(__APPLE__)
+    regex_t *preg_regex_c_1;
+#else
     struct re_pattern_buffer *preg_regex_c_1; /* from regex.c 339 */
+#endif
     int *regex_ok_regex_c_1; /* from regex.c 340 */
 
 /*
diff --git a/src/wave_view.c b/src/wave_view.c
index 7bd9f34..afe6bdf 100644
--- a/src/wave_view.c
+++ b/src/wave_view.c
@@ -685,6 +685,7 @@ static gboolean gw_wave_view_draw(GtkWidget *widget, cairo_t *cr)
 static void gw_wave_view_size_allocate(GtkWidget *widget, GtkAllocation *allocation)
 {
     GwWaveView *self = GW_WAVE_VIEW(widget);
+    int scale;
 
     GTK_WIDGET_CLASS(gw_wave_view_parent_class)->size_allocate(widget, allocation);
 
@@ -698,12 +699,14 @@ static void gw_wave_view_size_allocate(GtkWidget *widget, GtkAllocation *allocat
 
     g_clear_pointer(&self->traces_surface, cairo_surface_destroy);
 
+    scale = gtk_widget_get_scale_factor(widget);
+
     self->traces_surface =
         gdk_window_create_similar_image_surface(gtk_widget_get_window(widget),
                                                 CAIRO_FORMAT_ARGB32,
-                                                allocation->width,
-                                                allocation->height,
-                                                gtk_widget_get_scale_factor(widget));
+                                                allocation->width * scale,
+                                                allocation->height * scale,
+                                                scale);
 
     self->dirty = TRUE;
 }
-- 
2.39.3 (Apple Git-145)

